src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs

```diff
@@ -236,34 +230,26 @@ public virtual void Write(char[] chars, int index, int count)
         // Writes a double to this stream. The current position of the stream is
         // advanced by eight.
         //
-        public virtual unsafe void Write(double value)
+        public virtual void Write(double value)
         {
-            ulong TmpValue = *(ulong*)&value;
-            _buffer[0] = (byte)TmpValue;
-            _buffer[1] = (byte)(TmpValue >> 8);
-            _buffer[2] = (byte)(TmpValue >> 16);
-            _buffer[3] = (byte)(TmpValue >> 24);
-            _buffer[4] = (byte)(TmpValue >> 32);
-            _buffer[5] = (byte)(TmpValue >> 40);
-            _buffer[6] = (byte)(TmpValue >> 48);
-            _buffer[7] = (byte)(TmpValue >> 56);
-            OutStream.Write(_buffer, 0, 8);
+            Write(BitConverter.DoubleToInt64Bits(value));
         }
 
         public virtual void Write(decimal value)
         {
-            decimal.GetBytes(value, _buffer);
-            OutStream.Write(_buffer, 0, 16);
+            Span<byte> span = stackalloc byte[sizeof(decimal)];
+            decimal.GetBytes(in value, span);
+            OutStream.Write(span);
         }
 
         // Writes a two-byte signed integer to this stream. The current position of
         // the stream is advanced by two.
         //
-        public virtual void Write(short value)
+        public virtual unsafe void Write(short value)
         {
-            _buffer[0] = (byte)value;
-            _buffer[1] = (byte)(value >> 8);
-            OutStream.Write(_buffer, 0, 2);
+            if (!BitConverter.IsLittleEndian)
+                value = BinaryPrimitives.ReverseEndianness(value);
+            OutStream.Write(new ReadOnlySpan<byte>(&value, sizeof(short)));
```

Comment:
`Stream.Write` method that takes `Span` is not overridden in many older Stream implementations. This change will cause significant performance regression when use BinaryWriter is used with one these older Streams.

Reply:
I acknowledge that performance is important, but we are entering a new framework version. Is legacy code performance _that_ important? Or maybe a small allocation of 16 bytes is not that significant whatsoever.

If it's better for the project, I will have no problem with closing this PR.

Reply:
This change is improving performance for some workloads, and regressing performance for other workloads. We do require performance measurements for any performance related changes. Then we decide based on the data whether the change is an improvement on average.

My gut-feel is that the data will tell us that this change is not an improvement on average. BinaryWriter is typically used in legacy code. I do not expect it being used in modern code that much.

So unless you feel strongly about this, it is probably best to close this PR.

Thank you for submitting the PR anyway!

Reply:
OK, you are welcome.